### ISSUE TITLE
Force ANSI escape sequences on windows

### ISSUE THREAD

**Issue by melund** [jonathanslenders/python-prompt-toolkit#340]
The preview version of windows 10 has added support for ANSI escapes sequences. Still only 16 colors, but 256 are coming.

Are there a way to force prompt_toolkit to use VT100 and ansi escapes instead of the API calls when running on windows?

**Comment by jonathanslenders** [comment]
Hi @melund,

Is there a reason you want to avoid any API calls at all?

Because there is no escape sequence for getting the terminal size, and it also doesn't support the CPR escape sequence. Would it be ok for you to mix them?

Right now, there is an output class ConEmuOutput that does exactly this: https://github.com/jonathanslenders/python-prompt-toolkit/blob/master/prompt_toolkit/terminal/conemu_output.py
This enables 256 colors in ConEmu. I suppose we could do the same for the Windows Console.

Do you know how to detect the capabilities of the Windows Console? Probably, at some point, I'd like to make ANSI 256color the default, but it should automatically fall back when this capability is not available.

Jonathan

**Comment by melund** [comment]
> Is there a reason you want to avoid any API calls at all?

Well. If I want to use the application through a SSH connection or similar. Otherwise, I dont see the problem with the API calls. I basically wanted to know if there was a way I could test the capabilities of ansi support in windows 10 Conhost. 

Does the API calls slow down prompt_toolkit? It seems that xonsh with prompt_toolkit is quite a bit slower on windows compared to linux. 

> Because there is no escape sequence for getting the terminal size, and it also doesn't support the CPR escape sequence. Would it be ok for you to mix them?

I don't see why not. But maybe these missing features will show up at some point. It seems that it is still a work in progress. 

> Do you know how to detect the capabilities of the Windows Console? Probably, at some point, I'd like to make ANSI 256color the default, but it should automatically fall back when this capability is not available. 

At the moment it just ignores 256color codes, and automatically find the nearest color of the 16 colors. 

How would I force prompt_toolkit in xonsh to use the ConEmuOutput class? 

/Morten

**Comment by jonathanslenders** [comment]
For now, I wouldn't use the ConEmuOutput class, unless it's for testing. I'm not confident enough to say that it won't give any other issues.

Regarding #333, this could speed up the start-up time, but won't change anything at runtime. To improve the rendering time, I created this issue: https://github.com/jonathanslenders/python-prompt-toolkit/issues/241

**Comment by melund** [comment]
@jonathanslenders 

The WSL team has updated the Windows console to support 24bit colors: https://blogs.msdn.microsoft.com/commandline/2016/09/22/24-bit-color-in-the-windows-console/

Someone asked when other programs would start to use the more than 16 colors. Here is reply:
> When those tools start emitting the necessary VT sequences to render > 16 colors. We don’t currently plan on updating the console API’s to support the many new features we have added, and continue to add, to support VT.

So if they won't update the API, PTK really should use VT sequences solely on Windows.

**Comment by melund** [comment]
@jonathanslenders. So it seems the console is able to handle something like Tmux only using VT escapes. 

https://blogs.msdn.microsoft.com/commandline/2016/06/08/tmux-support-arrives-for-bash-on-ubuntu-on-windows/

So wouldn't it be possible to run prompt-toolkit with the VT100 backend on windows?

**Comment by melund** [comment]
@jonathanslenders 

> Do you know how to detect the capabilities of the Windows Console? Probably, at some point, I'd like to make ANSI 256color the default, but it should automatically fall back when this capability is not available.

So I looked into this. You should be able to detect the capabilities of the console, by setting the `ENABLE_VIRTUAL_TERMINAL_PROCESSING` (0x04) on a screen buffer handle. 

Something like:
```
windll.kernel32.SetConsoleMode(CONSOLE_SCREEN_BUFFER_INFO, ENABLE_VIRTUAL_TERMINAL_PROCESSING)
```

If that fails, you know that the console doesn't support VT sequences. See here: https://github.com/Microsoft/BashOnWindows/issues/1356#issuecomment-260704549

**Comment by jonathanslenders** [comment]
Hi @melund,

This is implemented for prompt_toolkit 2.0. We now use ANSI sequences on Windows versions that support it. If ``PROMPT_TOOLKIT_COLOR_DEPTH=DEPTH_24_BIT`` has been set, we even support true color.

Thanks for the help!

**Comment by melund** [comment]
Thanks a lot @jonathanslenders. This is fantastic and it will likely to speed up things running on windows. I really hope we can soon move xonsh to PTK2.
